core/src/WsTrust/MexParser.cs

```diff
@@ -0,0 +1,280 @@
+
+        public Uri Url { get; set; }
+    }
+
+    internal static class MexParser
+    {
+        private const string WsTrustSoapTransport = "http://schemas.xmlsoap.org/soap/http";
+
+        public static async Task<WsTrustAddress> FetchWsTrustAddressFromMexAsync(string federationMetadataUrl, UserAuthType userAuthType, RequestContext requestContext)
+        // This method would return null if the wanted policy is not found in mexDocument. It could potentially throw XmlException or http-relevant exceptions too.
+        {
+            XDocument mexDocument = await FetchMexAsync(federationMetadataUrl, requestContext).ConfigureAwait(false);
+            return ExtractWsTrustAddressFromMex(mexDocument, userAuthType, requestContext);
+        }
+
+        internal static async Task<XDocument> FetchMexAsync(string federationMetadataUrl, RequestContext requestContext)
+        {
+            var httpResponse = await HttpRequest.SendGet( // It would potentially throw MsalServiceException
+                new UriBuilder(federationMetadataUrl).Uri, null, requestContext).ConfigureAwait(false);
+            if (httpResponse.StatusCode != System.Net.HttpStatusCode.OK) {
+                throw new Client.MsalException(
+                    MsalError.AccessingWsMetadataExchangeFailed,
+                    string.Format(CultureInfo.CurrentCulture,
+                        "Response status code does not indicate success: {0} ({1}).",
+                        (int)httpResponse.StatusCode, httpResponse.StatusCode));
+            }
+            return XDocument.Parse(httpResponse.Body, LoadOptions.None);
+        }
+
+        internal static WsTrustAddress ExtractWsTrustAddressFromMex(XDocument mexDocument, UserAuthType userAuthType, RequestContext requestContext)
+        // This method would return null if the wanted policy is not found in mexDocument. It could potentially throw XmlException.
+        {
+            WsTrustAddress address = null;
+            MexPolicy policy = null;
+
+            Dictionary<string, MexPolicy> policies = ReadPolicies(mexDocument);
+            Dictionary<string, MexPolicy> bindings = ReadPolicyBindings(mexDocument, policies);
+            SetPolicyEndpointAddresses(mexDocument, bindings);
+            Random random = new Random();
+            //try ws-trust 1.3 first
+            policy = policies.Values.Where(p => p.Url != null && p.AuthType == userAuthType && p.Version == WsTrustVersion.WsTrust13).OrderBy(p => random.Next()).FirstOrDefault() ??
+                        policies.Values.Where(p => p.Url != null && p.AuthType == userAuthType).OrderBy(p => random.Next()).FirstOrDefault();
+
+            if (policy != null)
+            {
+                address = new WsTrustAddress();
+                address.Uri = policy.Url;
+                address.Version = policy.Version;
+            }
+            return address;
+        }
+
+        internal static Dictionary<string, MexPolicy> ReadPolicies(XContainer mexDocument)
```

Comment:
ReadMexPolicies ?

Reply:
Since this ReadPolicies(...) method is already inside a class named MexParser, I don't feel a strong need to rename it to ReadMexPolicies(...). But if you really insist to do so, let me know and I'll change it.

---
In reply to: [201820108](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r201820108) [](ancestors = 201820108)